cbits/x86/is-valid-utf8.c

```diff
@@ -0,0 +1,707 @@
+    0x7F, 0xBF, 0xBF, 0xBF, 0xBF, 0x9F, 0xBF, 0x8F, 0xF4, 0x80, 0x80,
+    0x80, 0x80, 0x80, 0x80, 0x80, 0x7F, 0xBF, 0xBF, 0xBF, 0xBF, 0x9F,
+    0xBF, 0x8F, 0xF4, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80,
+};
+
+static int8_t const df_ee_lookup2[32] = {
+    0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0,
+    0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0,
+};
+
+static int8_t const ef_fe_lookup2[32] = {
+    0, 3, 0, 0, 0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 3, 0, 0, 0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+};
+
+__attribute__((target("avx,avx2"))) static inline __m256i
+high_nibbles_of_avx2(__m256i const src) {
+  return _mm256_and_si256(_mm256_srli_epi16(src, 4), _mm256_set1_epi8(0x0F));
+}
+
+__attribute__((target("avx,avx2"))) static inline __m256i
+push_last_byte_of_a_to_b(__m256i const a, __m256i const b) {
+  return _mm256_alignr_epi8(b, _mm256_permute2x128_si256(a, b, 0x21), 15);
+}
+
+__attribute__((target("avx,avx2"))) static inline __m256i
+push_last_2bytes_of_a_to_b(__m256i const a, __m256i const b) {
+  return _mm256_alignr_epi8(b, _mm256_permute2x128_si256(a, b, 0x21), 14);
+}
+
+__attribute__((target("avx,avx2"))) static inline __m256i
+push_last_3bytes_of_a_to_b(__m256i const a, __m256i const b) {
+  return _mm256_alignr_epi8(b, _mm256_permute2x128_si256(a, b, 0x21), 13);
+}
+
+__attribute__((target("avx,avx2"))) static inline void
+check_block_avx2(__m256i const prev_input, __m256i const prev_first_len,
+                 __m256i *errors, __m256i const first_range_tbl,
+                 __m256i const range_min_tbl, __m256i const range_max_tbl,
+                 __m256i const df_ee_tbl, __m256i const ef_fe_tbl,
+                 __m256i const input, __m256i const first_len) {
+  __m256i range =
+      _mm256_shuffle_epi8(first_range_tbl, high_nibbles_of_avx2(input));
+  range = _mm256_or_si256(range,
+                          push_last_byte_of_a_to_b(prev_first_len, first_len));
+  __m256i tmp1 = push_last_2bytes_of_a_to_b(prev_first_len, first_len);
+  __m256i tmp2 = _mm256_subs_epu8(tmp1, _mm256_set1_epi8(0x01));
+  range = _mm256_or_si256(range, tmp2);
+  tmp1 = push_last_3bytes_of_a_to_b(prev_first_len, first_len);
+  tmp2 = _mm256_subs_epu8(tmp1, _mm256_set1_epi8(0x02));
+  range = _mm256_or_si256(range, tmp2);
+  __m256i const shift1 = push_last_byte_of_a_to_b(prev_input, input);
+  __m256i pos = _mm256_sub_epi8(shift1, _mm256_set1_epi8(0xEF));
+  tmp1 = _mm256_subs_epu8(pos, _mm256_set1_epi8(0xF0));
+  __m256i range2 = _mm256_shuffle_epi8(df_ee_tbl, tmp1);
+  tmp2 = _mm256_adds_epu8(pos, _mm256_set1_epi8(0x70));
+  range2 = _mm256_add_epi8(range2, _mm256_shuffle_epi8(ef_fe_tbl, tmp2));
+  range = _mm256_add_epi8(range, range2);
+  __m256i const minv = _mm256_shuffle_epi8(range_min_tbl, range);
+  __m256i const maxv = _mm256_shuffle_epi8(range_max_tbl, range);
+  errors[0] = _mm256_or_si256(errors[0], _mm256_cmpgt_epi8(minv, input));
+  errors[1] = _mm256_or_si256(errors[1], _mm256_cmpgt_epi8(input, maxv));
+}
```

Comment:
I rather think that code like this needs to be extensively commented.  The above looks like line-noise to me.

Reply:
This is identical to the code for the SSE version, which _is_ extensively commented. The only change is the use of AVX intrinsics.

Reply:
I rather think, It still deserves documentation, even if it is a cut/paste job.  The functions could diverge later, and any errors would be harder to notice if it is changed without in place comments explaining what this is doing.